### ISSUE TITLE
Strictness is a pain

### ISSUE THREAD

**Issue by scole66** [scole66/ecma262#221]
In Directive Prologue.

A Directive Prologue is the longest sequence of ExpressionStatements occurring as the initial StatementListItems or ModuleItems of a FunctionBody, a ScriptBody, or a ModuleBody and where each ExpressionStatement in the sequence consists entirely of a StringLiteral token followed by a semicolon. The semicolon may appear explicitly or may be inserted by automatic semicolon insertion. A Directive Prologue may be an empty sequence.

Current:
```python
def parse_StatementList(ctx, lexer, Yield, Await, Return):
    # 13.2 Block
    # Syntax
    #   StatementList[Yield, Await, Return] :
    #       StatementListItem[?Yield, ?Await, ?Return]
    #       StatementList[?Yield, ?Await, ?Return] StatementListItem[?Yield, ?Await, ?Return]
    def parse(collected=None):
        sli = parse_StatementListItem(ctx, lexer, Yield, Await, Return)
        if not sli:
            return collected
        if collected is None:
            return lambda: parse(P2_StatementList_StatementListItem(ctx, [sli]))
        # Tail recursion replaced by a thunk; please use trampoline.
        return lambda: parse(P2_StatementList_StatementList_StatementListItem(ctx, [collected, sli]))

    return trampoline(parse)
```

So what do we do here?

Maybe this:
```python
def parse_statementlist(dp_active, strict):
    def parse(collected=None):
        sli = parse_StatementListItem(strict)
        if not sli:
            return collected
        dp_active = dp_active and sli.IsStringLiteral
        use_strict = dp_active and sli.HasUseStrict
        strict = strict or use_strict
        if collected is None:
            return lambda: parse(P2_StatementList_StatementListItem(ctx, strict, [sli]))
        return lambda: parse(P2_StatementList_StatementList_StatementListItem(ctx, strict, [collected, sli]))
    return trampoline(parse)
```
That means a kind of large change to the parse_StatementList API; and adding "strict" to every other parse function. I don't want to store it in "context" because if it's started inside a function, it shouldn't pollute outside of function code.

**Comment by scole66** [comment]
Need to make sure that dp_active propogates through the recursive calls to parse. I _think_ it does, but I'm not sure.

**Comment by scole66** [comment]
Fixed in #223.
